File: libs/sdk-python/src/daytona_sdk/_async/object_storage.py
# Copyright 2025 Daytona Platforms Inc.
# SPDX-License-Identifier: AGPL-3.0

import asyncio
import hashlib
import os
import tarfile
import threading

import aioboto3

from .._utils.docs_ignore import docs_ignore


class ObjectStorage:
    """ObjectStorage class for interacting with object storage services.

    Attributes:
        endpoint_url (str): The endpoint URL for the object storage service.
        aws_access_key_id (str): The access key ID for the object storage service.
        aws_secret_access_key (str): The secret access key for the object storage service.
        aws_session_token (str): The session token for the object storage service. Used for temporary credentials.
        bucket_name (str): The name of the bucket to use.
    """

    def __init__(
        self,
        endpoint_url,
        aws_access_key_id,
        aws_secret_access_key,
        aws_session_token,
        bucket_name="daytona-volume-builds",
    ):
        self.bucket_name = bucket_name
        self.endpoint_url = endpoint_url
        self.s3_client = aioboto3.Session().client(
            service_name="s3",
            endpoint_url=self.endpoint_url,
            aws_access_key_id=aws_access_key_id,
            aws_secret_access_key=aws_secret_access_key,
            aws_session_token=aws_session_token,
        )
        # unasync: delete start
        self._client_ctx = None

    async def open_client(self):
        try:
            self._client_ctx = self.s3_client
            # pylint: disable=unnecessary-dunder-call
            self.s3_client = await self._client_ctx.__aenter__()
        except Exception as e:
            raise Exception(f"Error opening S3 client: {e}") from e

    async def close_client(self):
        try:
            await self._client_ctx.__aexit__(None, None, None)
            self.s3_client = None
            self._client_ctx = None
        except Exception as e:
            raise Exception(f"Error closing S3 client: {e}") from e

    # unasync: delete end

    async def upload(self, path, organization_id, archive_base_path=None) -> str:
        """Uploads a file to the object storage service.

        Args:
            path (str): The path to the file to upload.
            organization_id (str): The organization ID to use.
            archive_base_path (str): The base path to use for the archive.
        """
        if not os.path.exists(path):
            raise FileNotFoundError(f"Path does not exist: {path}")

        # Compute hash for the path
        path_hash = self._compute_hash_for_path_md5(path, archive_base_path)

        # Define the S3 prefix
        prefix = f"{organization_id}/{path_hash}/"
        s3_key = f"{prefix}context.tar"

        # Check if it already exists in S3
        if await self._folder_exists_in_s3(prefix):
            return path_hash

        # Upload to S3
        await self._upload_as_tar(s3_key, path, archive_base_path)

        return path_hash

    @staticmethod
    @docs_ignore
    def compute_archive_base_path(path_str) -> str:
        """Compute the base path for an archive. Returns normalized path without the root
        (drive letter or leading slash).

        Args:
            path_str (str): The path to compute the base path for.

        Returns:
            str: The base path for the given path.
        """
        path_str = os.path.normpath(path_str)
        # Remove drive letter for Windows paths (e.g., C:)
        _, path_without_drive = os.path.splitdrive(path_str)
        # Remove leading separators (both / and \)
        return path_without_drive.lstrip("/").lstrip("\\")

    def _compute_hash_for_path_md5(self, path_str, archive_base_path=None):
        """Computes the MD5 hash for a given path.

        Args:
            path_str (str): The path to compute the hash for.
            archive_base_path (str): The base path to use for the archive.

        Returns:
            str: The MD5 hash for the given path.
        """
        md5_hasher = hashlib.md5()
        abs_path_str = os.path.abspath(path_str)

        if archive_base_path is None:
            archive_base_path = self.compute_archive_base_path(path_str)
        md5_hasher.update(archive_base_path.encode("utf-8"))

        if os.path.isfile(abs_path_str):
            with open(abs_path_str, "rb") as f:
                for chunk in iter(lambda: f.read(8192), b""):
                    md5_hasher.update(chunk)
        else:
            for root, dirs, files in os.walk(abs_path_str):
                if not dirs and not files:
                    rel_dir = os.path.relpath(root, path_str)
                    md5_hasher.update(rel_dir.encode("utf-8"))
                for filename in files:
                    file_path = os.path.join(root, filename)
                    rel_path = os.path.relpath(file_path, abs_path_str)

                    # Incorporate the relative path
                    md5_hasher.update(rel_path.encode("utf-8"))

                    # Incorporate file contents
                    with open(file_path, "rb") as f:
                        for chunk in iter(lambda: f.read(8192), b""):
                            md5_hasher.update(chunk)

        return md5_hasher.hexdigest()

    async def _folder_exists_in_s3(self, prefix):
        """Checks if a folder exists in S3.

        Args:
            prefix (str): The prefix to check.

        Returns:
            bool: True if the folder exists, False otherwise.
        """
        resp = await self.s3_client.list_objects_v2(Bucket=self.bucket_name, Prefix=prefix)
        return "Contents" in resp

    async def _upload_as_tar(self, s3_key, source_path, archive_base_path=None):
        """Uploads a file to the object storage service as a tar.

        Args:
            s3_key (str): The key to upload the file to.
            source_path (str): The path to the file to upload.
            archive_base_path (str): The base path to use for the archive.
        """
        source_path = os.path.normpath(source_path)

        if archive_base_path is None:
            archive_base_path = self.compute_archive_base_path(source_path)

        read_fd, write_fd = os.pipe()
        read_file = os.fdopen(read_fd, "rb")
        write_file = os.fdopen(write_fd, "wb")

        def tar_worker():
            with tarfile.open(fileobj=write_file, mode="w|") as tar:
                tar.add(source_path, arcname=archive_base_path)
            write_file.close()

        thread = threading.Thread(target=tar_worker, daemon=True)
        thread.start()

        await self.s3_client.upload_fileobj(Fileobj=read_file, Bucket=self.bucket_name, Key=s3_key)

        read_file.close()
        await asyncio.to_thread(thread.join)

File: libs/sdk-typescript/src/ObjectStorage.ts
/*
 * Copyright 2025 Daytona Platforms Inc.
 * SPDX-License-Identifier: Apache-2.0
 */

import { ListObjectsV2Command, S3Client } from '@aws-sdk/client-s3'
import { Upload } from '@aws-sdk/lib-storage'
import * as crypto from 'crypto'
import * as pathe from 'pathe'
import { DaytonaError } from './errors/DaytonaError'
import { dynamicImport } from './utils/Import'

/**
 * Configuration for the ObjectStorage class.
 *
 * @interface
 * @property {string} endpointUrl - The endpoint URL for the object storage service.
 * @property {string} accessKeyId - The access key ID for the object storage service.
 * @property {string} secretAccessKey - The secret access key for the object storage service.
 * @property {string} [sessionToken] - The session token for the object storage service. Used for temporary credentials.
 * @property {string} [bucketName] - The name of the bucket to use.
 */
export interface ObjectStorageConfig {
  endpointUrl: string
  accessKeyId: string
  secretAccessKey: string
  sessionToken?: string
  bucketName?: string
}

/**
 * ObjectStorage class for interacting with object storage services.
 *
 * @class
 * @param {ObjectStorageConfig} config - The configuration for the object storage service.
 */
export class ObjectStorage {
  private bucketName: string
  private s3Client: S3Client

  constructor(config: ObjectStorageConfig) {
    this.bucketName = config.bucketName || 'daytona-volume-builds'
    this.s3Client = new S3Client({
      region: this.extractAwsRegion(config.endpointUrl) || 'us-east-1',
      endpoint: config.endpointUrl,
      credentials: {
        accessKeyId: config.accessKeyId,
        secretAccessKey: config.secretAccessKey,
        sessionToken: config.sessionToken,
      },
      forcePathStyle: true,
    })
  }

  /**
   * Upload a file or directory to object storage.
   *
   * @param {string} path - The path to the file or directory to upload.
   * @param {string} organizationId - The organization ID to use for the upload.
   * @param {string} archiveBasePath - The base path to use for the archive.
   * @returns {Promise<string>} The hash of the uploaded file or directory.
   */
  async upload(path: string, organizationId: string, archiveBasePath: string): Promise<string> {
    const fs = await dynamicImport('fs', '"upload" is not supported: ')

    if (!fs.existsSync(path)) {
      const errMsg = `Path does not exist: ${path}`
      throw new DaytonaError(errMsg)
    }

    // Compute hash for the path
    const pathHash = await this.computeHashForPathMd5(path, archiveBasePath)

    // Define the S3 prefix
    const prefix = `${organizationId}/${pathHash}/`
    const s3Key = `${prefix}context.tar`

    // Check if it already exists in S3
    if (await this.folderExistsInS3(prefix)) {
      return pathHash
    }

    // Upload to S3
    await this.uploadAsTar(s3Key, path, archiveBasePath)

    return pathHash
  }

  /**
   * Compute a hash for a file or directory.
   *
   * @param {string} pathStr - The path to the file or directory to hash.
   * @param {string} archiveBasePath - The base path to use for the archive.
   * @returns {Promise<string>} The hash of the file or directory.
   */
  private async computeHashForPathMd5(pathStr: string, archiveBasePath: string): Promise<string> {
    const fs = await dynamicImport('fs', '"computeHashForPathMd5" is not supported: ')

    const md5Hasher = crypto.createHash('md5')
    const absPathStr = pathe.resolve(pathStr)

    md5Hasher.update(archiveBasePath)

    if (fs.statSync(absPathStr).isFile()) {
      // For files, hash the content
      await this.hashFile(absPathStr, md5Hasher)
    } else {
      // For directories, recursively hash all files and their paths
      await this.hashDirectory(absPathStr, pathStr, md5Hasher)
    }

    return md5Hasher.digest('hex')
  }

  /**
   * Recursively hash a directory and its contents.
   *
   * @param {string} dirPath - The path to the directory to hash.
   * @param {string} basePath - The base path to use for the hash.
   * @param {crypto.Hash} hasher - The hasher to use for the hash.
   * @returns {Promise<void>} A promise that resolves when the directory has been hashed.
   */
  private async hashDirectory(dirPath: string, basePath: string, hasher: crypto.Hash): Promise<void> {
    const fs = await dynamicImport('fs', '"hashDirectory" is not supported: ')

    const entries = fs.readdirSync(dirPath, { withFileTypes: true })
    const hasSubdirs = entries.some((e) => e.isDirectory())
    const hasFiles = entries.some((e) => e.isFile())

    if (!hasSubdirs && !hasFiles) {
      // Empty directory
      const relDir = pathe.relative(basePath, dirPath)
      hasher.update(relDir)
    }

    for (const entry of entries) {
      const fullPath = pathe.join(dirPath, entry.name)

      if (entry.isDirectory()) {
        await this.hashDirectory(fullPath, basePath, hasher)
      } else if (entry.isFile()) {
        const relPath = pathe.relative(basePath, fullPath)
        hasher.update(relPath)

        await this.hashFile(fullPath, hasher)
      }
    }
  }

  /**
   * Hash a file.
   *
   * @param {string} filePath - The path to the file to hash.
   * @param {crypto.Hash} hasher - The hasher to use for the hash.
   * @returns {Promise<void>} A promise that resolves when the file has been hashed.
   */
  private async hashFile(filePath: string, hasher: crypto.Hash): Promise<void> {
    const fs = await dynamicImport('fs', '"hashFile" is not supported: ')

    await new Promise<void>((resolve, reject) => {
      const stream = fs.createReadStream(filePath, { highWaterMark: 8192 })
      stream.on('data', (chunk) => hasher.update(chunk))
      stream.on('end', resolve)
      stream.on('error', reject)
    })
  }

  /**
   * Check if a prefix (folder) exists in S3.
   *
   * @param {string} prefix - The prefix to check.
   * @returns {Promise<boolean>} True if the prefix exists, false otherwise.
   */
  private async folderExistsInS3(prefix: string): Promise<boolean> {
    const response = await this.s3Client.send(
      new ListObjectsV2Command({
        Bucket: this.bucketName,
        Prefix: prefix,
        MaxKeys: 1,
      }),
    )

    return !!response.Contents && response.Contents.length > 0
  }

  /**
   * Create a tar archive of the specified path and upload it to S3.
   *
   * @param {string} s3Key - The key to use for the uploaded file.
   * @param {string} sourcePath - The path to the file or directory to upload.
   * @param {string} archiveBasePath - The base path to use for the archive.
   */
  private async uploadAsTar(s3Key: string, sourcePath: string, archiveBasePath: string) {
    const importErrorPrefix = '"uploadAsTar" is not supported: '
    const tar = await dynamicImport('tar', importErrorPrefix)
    const stream = await dynamicImport('stream', importErrorPrefix)

    sourcePath = pathe.resolve(sourcePath)
    const normalizedSourcePath = pathe.normalize(sourcePath)
    const normalizedArchiveBasePath = pathe.normalize(archiveBasePath)

    let basePrefix: string

    if (normalizedArchiveBasePath === '.') {
      // When archiveBasePath is empty (normalized to '.'), use the normalizedSourcePath as cwd and the '.' as target
      basePrefix = normalizedSourcePath
    } else {
      // Normal case: extract the base prefix by removing archiveBasePath from the end
      basePrefix = normalizedSourcePath.slice(0, normalizedSourcePath.length - normalizedArchiveBasePath.length)
    }

    const tarStream = tar.create(
      {
        cwd: basePrefix,
        portable: true,
        gzip: false,
      },
      [normalizedArchiveBasePath],
    )

    const pass = new stream.PassThrough()
    tarStream.pipe(pass)

    const uploader = new Upload({
      client: this.s3Client,
      params: {
        Bucket: this.bucketName,
        Key: s3Key,
        Body: pass,
      },
    })

    await uploader.done()
  }

  private extractAwsRegion(endpoint: string): string | undefined {
    const match = endpoint.match(/s3[.-]([a-z0-9-]+)\.amazonaws\.com/)
    return match?.[1]
  }
}
